### ISSUE TITLE
Disable quick connect until a condition is met

### ISSUE THREAD

**Issue by akhil-sq** [amazon-connect/amazon-connect-streams#708]
Hey team,
We use the custom ccp as an app for our contact centers. We heavily use quick connects for transfers. However, we have a new requirement where we want to disable the DOM for quick connect until the agent has filled the transfer summary in a separate modal. 
Since the containerDiv is an iframe and we don't have access to the DOM I believe, but is there a way we can access this DOM and disable it and then enable it again?

**Comment by hxgnZO** [comment]
Depends on your implementation/hosting solution I would think. My first inclination is to set the agent's permissions dynamically- I believe the QC being displayed in the CCP is an init param- but the connect core still runs in background- so perhaps you re-init the CCP every time you need to hide it but this seems really redundant. I have had to use some DOM manipulation/janky workarounds in our solution due to how it is hosted- but I really wish I had the time to build a custom CCP, sounds like that might be a good option for you as well. You might be able to set a requirement for some sort of flag in the react app without going too deep into it- but iframes definitely make any sort of front end stuff annoying- IE: agent filled out a call transfer form on front end- trying to make the ccp understand that something happened. The front end layer that binds the two just isn't sufficient in my experience, and my understanding of react isn't quite good enough to rebuild the CCP how I would like it. 
Anyone who has been deeper into the connect resources on the backend might be able to provide some insight- there's probably a Lambda or DDB action or something that is involved, you may be able to leverage that with some sort of custom event that updates whatever resource is being used to make quick connects available

**Comment by hxgnZO** [comment]
This little bit of the core functionality might be leveraged to do what you're talking about. Specifically the client.call(connect.ClientMethods.{whichever methods are avail})


    var client = connect.core.getClient();
  //................//
      client.call(connect.ClientMethods.UPDATE_AGENT_CONFIGURATION, {
        configuration: connect.assertNotNull(configuration, 'configuration')
      }, {
          success: function (data) {
            // We need to ask the shared worker to reload agent config
            // once we change it so every tab has accurate config.
            var conduit = connect.core.getUpstream();
            conduit.sendUpstream(connect.EventType.RELOAD_AGENT_CONFIGURATION);
//.............//

**Comment by akhil-sq** [comment]
Thanks for sharing! I don't know how the reloading would effect the advocate experience. In theory, I want a connect stream event that I can use to block quick connect initially and then unblock it in my js code. Is that possible from connect streams?

**Comment by hxgnZO** [comment]
> Thanks for sharing! I don't know how the reloading would effect the advocate experience. In theory, I want a connect stream event that I can use to block quick connect initially and then unblock it in my js code. Is that possible from connect streams?

Not that I'm aware of using only the built in API, you would need to customize the React app before build to do this effectively I believe- and then need to be able to know which quick connects to enable. However, if you are using the agent experience thing and workflows in AWS, I'm sure it can be done through a lambda/in the contact flows. In fact, may still be possible to do in a contact flow, but that gets tricky if your required modal is in a separate app/workflow. 
We have some situations where I also would like to be able to trigger certain API aspects based on an external event. Doing it via DOM manipulation is front end suicide- the react app changes too much. What you could try as mentioned is on an event- change the agent options through the API. But I have not yet looked for something like this. If you have the ability to track which contacts you want made available to transfer by phone number in your hosting app/outside page then it could be done that way- event goes to outside page- which is able to retrieve the contact to transfer to somehow- then the outside page uses the streams API to transfer call to a specific number/contact programmatically.

I would think one way though, is to use a contact flow verification that puts the customer on hold and looks for a certain condition to be met. It would require some sort of integration I'm sure such that the flow is able to ascertain via Lambda or the like that the required agent work is completed. Really though, this sort of integration needs a custom integration of streams into whatever application the agent's workflow is in. In fact I dearly wish I had the time to build an API and the custom application to do just that for our help desk folks. A separate window entirely for the phone that is essentially a "desktop" app that will always ring and be stable regardless of other browser activity, but is somehow integrated into their workflows. I am pretty close to something like that for chat, but moving it into voice is slightly different.

I would just recommend getting into the resources provided in streams library, and learning how (if like me, it is new to you) the react application is built, then you could pretty easily pick through and add to the transfer function a requirement that is looking for some property on the outside page, maybe temp storage somewhere? A "ready to transfer" property, that is only a value when they are in a workflow where they must complete something, but otherwise undefined or something.

**Comment by hngan** [comment]
We don't take feature requests in github. Please contact the official Amazon Connect support and have them file a ticket to us.https://aws.amazon.com/contact-us/?nc2=h_header
